client/lib/route/path.js

```diff
@@ -63,12 +63,19 @@ function addSiteFragment( path, site ) {
 	return pieces.join( '/' );
 }
 
-function sectionify( path ) {
+function sectionify( path, siteFragment ) {
 	let basePath = path.split( '?' )[ 0 ];
-	const site = getSiteFragment( basePath );
 
-	if ( site ) {
-		basePath = trailingslashit( basePath ).replace( '/' + site + '/', '/' );
+	// Sometimes the caller knows better than `getSiteFragment` what the `siteFragment` is.
+	// For example, when the `:site` parameter is not the last or second-last part of the route
+	// and is retrieved from `context.params.site`. In that case, it can pass the `siteFragment`
+	// explicitly as the second parameter. We call `getSiteFragment` only as a fallback.
+	if ( ! siteFragment ) {
+		siteFragment = getSiteFragment( basePath );
```

Comment:
Hmm, I'm not familiar enough with how `sectionify` is called across the codebase to have an opinion about a change like this.

In the `context.params` object, the site ID property is not always named `site`. Often it's `site_id`. At other places, it's `siteOrCategory` or even `plugin` :wink: (the more funny names are used in ambiguous routes like `/plugins/woocommerce` vs `/plugins/jsnajdr.blog`). The `sectionify` function wouldn't know where to look without additional info.